FuelLabs/fuelup
# feat: allow usage of fuel-toolchain.toml in projects

closes #65 

## Summary

This should be good to go for review. Basically this PR implements
- the store,
- basic overriding with `channel` and `components`.
- displaying the overridden toolchain with `fuelup default`.

### How does this work?
Whenever we execute binaries installed via fuelup, we always check in `proxy_cli.rs` the existence of an override file in the project root. From here, `fuelup` will decide which toolchain or component to use, and decide to install components (if not installed).

To test this PR:
1. checkout this PR, `cargo build`
2. link the built fuelup and a chosen fuelup binary of your choice to test. For me I am using `forc` and have this script to quickly iterate/test things:

```sh
rm -f ~/.fuelup/bin/forc
ln target/debug/fuelup ~/.fuelup/bin/forc
```

3. create a `fuel-toolchain.toml` at your current working directory, edit it to your wish to test things. You can see the [tests](https://github.com/FuelLabs/fuelup/blob/10e2807e5ef2b7470932a35e109002e2009da87d/src/toolchain_override.rs#L144) for a few examples.


This PR does **NOT** yet support:
- specifying the dates (eg. `nightly-2022-01-09` won't work, and by extension the dated latest versions) https://github.com/FuelLabs/fuelup/issues/331
- installing to store via `fuelup component add`. Ultimately the goal is to move all downloads to the store and link components from there. https://github.com/FuelLabs/fuelup/issues/329
- `fuelup show` - currently it behaves like before, but we want to show if something is overriding. https://github.com/FuelLabs/fuelup/issues/330

The reason the above few features aren't supported in this PR is for reviewability. Keeping the diff small would make reviews easier. The above 3 features will be implemented in the next few PRs.


~Opening a draft PR to track the deliverables below:~

## Todos

- [x] fuelup should adhere to both the naming conventions of distributed toolchains and custom toolchains
- [x] fuelup should auto download the toolchain, or any missing components, when any `fuelup` managed binaries are invoked with this toml file in the project root
- [x] ~`fuelup show` should show the override as the currently active toolchain, instead of what is shown in `fuelup default`~ deferred to separate PR for reviewability
- [x] `fuelup default` should show current override as well
- [x] document `fuel-toolchain.toml` in the fuelup book
- [x] add some tests with `fuel-toolchain.toml`

Additional todos, following updates [here](https://github.com/FuelLabs/fuelup/issues/320):
- [x] download missing components to a new directory, `/.fuelup/store`, instead of in the toolchain context
- [x] execute components from the store directly instead
